File: 0143-reorder-list/0143-reorder-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
void reorderList(struct ListNode* head){
    if (head == NULL)
        return;
    if (head->next == NULL)
        return;
    if (head->next->next == NULL)
        return;
    
    struct ListNode *slow = head;
    struct ListNode *fast = head;
    
    while (fast != NULL && fast->next != NULL) {
        slow = slow->next;
        fast = fast->next->next;
    }
    
    struct ListNode *start_of_second_list = slow->next;
    slow->next = NULL;
    slow = start_of_second_list;
    struct ListNode *medium = slow->next;
    
    while (medium != NULL) {
        struct ListNode *next = medium->next;
        medium->next = slow;
        slow = medium;
        medium = next;
    }
    
    start_of_second_list->next = NULL;
    //head will be head of forwards list
    //slow will be head of backwards list
    while (head != NULL && slow != NULL) {
        struct ListNode *forwardNext = head->next;
        struct ListNode *backwardNext = slow->next;
        head->next = slow;
        slow->next = forwardNext;
        head = forwardNext;
        slow = backwardNext;
    }
}

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow, fast = head, head.next
        
        while (fast and fast.next):
            slow=slow.next
            fast=fast.next.next
            
        afterSlow = slow.next
        slow.next = None
        while afterSlow:
            temp = afterSlow.next
            afterSlow.next = slow
            slow = afterSlow
            afterSlow = temp
        
        while head and slow:
            temp1 = head.next
            temp2 = slow.next
            head.next = slow
            slow.next = temp1
            head = temp1
            slow = temp2
            